Allow ActorInits to target specific traits

pchote opened the pull request:
This PR finishes the work started in #18123 and #18164, fixing #18118.
See that issue for the rationale and an explanation of the syntax.

pchote commented:
This breaks a few things in the editor. Some will need to be fixed here, others can be deferred to a followup PR.

teinarss commented on OpenRA.Mods.Common/Scripting/Global/ActorGlobal.cs:
```diff
@@ -24,6 +27,39 @@ public class ActorGlobal : ScriptGlobal
+		ActorInit CreateInit(string initName, LuaValue value)
+		{
+			// Find the requested type
+			var initInstance = initName.Split('@');
```
IMO we should start using constants for magic strings and numbers.

pchote replied:
Done.

pchote commented:
Updated.

The editor can now correctly round-trip a map that uses multiple inits, but it will crash if you select an actor that uses them. We don't expose the multi-init stuff anywhere yet, so this is best left to the next PR which was already going to rework that logic.

pchote commented:
I'd like to do some further work here to make sure we can support composite init types (i.e. structs with multiple values), which we want for save state.

pchote committed:
TESTCASE

pchote commented:
Updated.

There is now a concept of "value" inits and "composite" inits, which are loaded from the yaml value and child nodes respectively. Composite inits are specified using a nested table from Lua.

The four chronoshift inits have been merged into a single `ChronoshiftReturnInit` to demonstrate how the composite inits work.

Looking forwards, I expect that each trait would define its own `*SerializedInit : CompositeInit` that holds all the state they need to persist over save/load.

pchote commented:
Fixed and rebased.

abcdefg30 commented on OpenRA.Game/Map/ActorReference.cs:
```diff
@@ -43,26 +45,43 @@ public ActorReference(string type, Dictionary<string, MiniYaml> inits)
+			foreach (var o in InitDict)
 			{
-				if (init is ISuppressInitExport)
+				if (!(o is ActorInit) || o is ISuppressInitExport)
```
Nit: Prefer `as` over `is` followed by a cast.

abcdefg30 commented on OpenRA.Game/Map/ActorReference.cs:
```diff
@@ -43,26 +45,43 @@ public ActorReference(string type, Dictionary<string, MiniYaml> inits)
+				var initTypeName = init.GetType().Name;
+				var initName = initTypeName.Substring(0, initTypeName.Length - 4);
+				if (!string.IsNullOrEmpty(init.InstanceName))
+					initName += "@" + init.InstanceName;
```
Can this use `ActorInfo.TraitInstanceSeparator` as well?

abcdefg30 commented on OpenRA.Mods.Common/Scripting/Global/ActorGlobal.cs:
```diff
@@ -24,6 +27,70 @@ public class ActorGlobal : ScriptGlobal
+				initType.GetField("InstanceName").SetValue(init, initInstance[1]);
+
+			var compositeInit = init as CompositeActorInit;
+			if (value is LuaTable && compositeInit != null)
```
Nit: Prefer `as` over `is` and a cast.

abcdefg30 commented on OpenRA.Mods.Common/ActorInitializer.cs:
```diff
@@ -10,77 +10,124 @@
-		[FieldFromYamlKey]
-		readonly Player value = null;
+		public readonly string InternalName;
+		Actor actor;
```
This can be `readonly` I think.

abcdefg30 commented on OpenRA.Mods.Common/ActorInitializer.cs:
```diff
@@ -10,77 +10,124 @@
+
+		public override object ConvertTo(ITypeDescriptorContext context, System.Globalization.CultureInfo culture, object value, Type destinationType)
+		{
+			if (destinationType == typeof(string) && value is ActorInitActorReference)
+				return ((ActorInitActorReference)value).InternalName;
```
Although a bit more verbose, the `as` convention strikes again:
```suggestion
			if (destinationType == typeof(string))
			{
				var reference = value as ActorInitActorReference;
				if (reference != null)
					return reference.InternalName;
			}
```

abcdefg30 commented on OpenRA.Game/Map/ActorInitializer.cs:
```diff
@@ -53,59 +65,175 @@ public ActorInitializer(Actor actor, TypeDictionary dict)
-		[FieldFromYamlKey]
-		readonly CPos value = CPos.Zero;
+		protected CompositeActorInit(TraitInfo info)
+			: base(info.InstanceName) { }
```
This crashes for me when killing a unit that is on the chronoshift countdown since `info` is null.

pchote committed:
Replace IActorInit with an abstract class.

A shared ValueActorInit<T> is introduced to reduce duplication
in the most common init cases, and an ActorInitActorReference
allow actors to be referenced by map.yaml name.

pchote committed:
Allow ActorInits to target a specific trait by matching the @ suffix.

pchote committed:
Add CompositeActorInit and simplify chronoshift inits.

pchote commented:
Fixed.

pchote commented:
This PR is a bottleneck for finishing the WAngle facing rework, with another 4 or 5 PRs still waiting on this one.

deleted-user-1 commented:
With [these](https://gist.githubusercontent.com/matjaeck/aeb16905928dff7838dbbff82f79b86c/raw/fabc5c343719ff9d75c68095550d02812e272e99/patch.diff) changes that have a copy/paste-error (`WithSpriteTurret@PRIMARY` defined 2x) the shellmap starts with the "Battlefield Control Initialized" notification. This is not the case on bleed - should the game crash instead?

Didn't spot obvious issues with AI games or saving/loading games.

pchote commented:
That is the standard behaviour when loading a shellmap with bad syntax - probably not ideal but nothing new.

deleted-user-1 reviewed (dismissed):
LGTM

abcdefg30 commented on OpenRA.Mods.Cnc/Traits/Chronoshiftable.cs:
```diff
@@ -59,14 +59,18 @@ public Chronoshiftable(ActorInitializer init, ChronoshiftableInfo info)
+				Origin = returnInit.Origin;
+
+				// Defer to the end of tick as the lazy value may reference an actor that hasn't been created yet
+				if (returnInit.Chronosphere != null)
```
This check is probably not needed as we will crash later on if no chronosphere is set. So we either remove it or throw an exception here. I tried using `FieldLoader.Require` on those properties but that didn't help either.

abcdefg30 commented on OpenRA.Mods.Cnc/Traits/Chronoshiftable.cs:
```diff
@@ -59,14 +59,18 @@ public Chronoshiftable(ActorInitializer init, ChronoshiftableInfo info)
+
+				// Defer to the end of tick as the lazy value may reference an actor that hasn't been created yet
+				if (returnInit.Chronosphere != null)
+					init.World.AddFrameEndTask(w => chronosphere = returnInit.Chronosphere.Actor(init.World).Value);
```
Something is still not working correctly here, I tried the following in allies 01 (just put it at the end of WorldLoaded):
```
	local cron = Actor.Create("PDOX", true, { Owner = player, Location = BaseCameraPoint.Location })
	Trigger.AfterDelay(2, function() Actor.Create("ca", true, { Owner = player, Location = Map.NamedActor("CruiserPoint2").Location, ChronoshiftReturn = { Origin = Map.NamedActor("CruiserPoint4").Location, Ticks = 100, Duration = 50, Chronosphere = cron }}) end)
```
and that crashes with
```
Exception of type `System.Collections.Generic.KeyNotFoundException`: Der angegebene Schlüssel war nicht im Wörterbuch angegeben.
   at System.ThrowHelper.ThrowKeyNotFoundException()
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at OpenRA.Mods.Common.ActorInitActorReference.InnerValue(World world) in OpenRA\OpenRA.Mods.Common\ActorInitializer.cs:Line 114.
   at OpenRA.Mods.Common.ActorInitActorReference.<>c__DisplayClass5_0.<Actor>b__0() in OpenRA\OpenRA.Mods.Common\ActorInitializer.cs:Line 124.
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.LazyInitValue()
   at OpenRA.Mods.Cnc.Traits.Chronoshiftable.<>c__DisplayClass7_1.<.ctor>b__0(World w) in OpenRA\OpenRA.Mods.Cnc\Traits\Chronoshiftable.cs:Line 72.
   at OpenRA.World.Tick() in OpenRA\OpenRA.Game\World.cs:Line 467.
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager) in OpenRA\OpenRA.Game\Game.cs:Line 602.
   at OpenRA.Game.LogicTick() in OpenRA\OpenRA.Game\Game.cs:Line 629.
   at OpenRA.Game.Loop() in OpenRA\OpenRA.Game\Game.cs:Line 793.
   at OpenRA.Game.Run() in OpenRA\OpenRA.Game\Game.cs:Line 834.
   at OpenRA.Game.InitializeAndRun(String[] args) in OpenRA\OpenRA.Game\Game.cs:Line 262.
   at OpenRA.Program.Main(String[] args) in OpenRA\OpenRA.Game\Support\Program.cs:Line 37.
```
as apparently the actor reference init is looking for a map placed actor (why?) instead of the actor I gave it.
When I don't give a `Chronosphere` argument, it doesn't crash when creating the init, but instead when the timer runs out and it tries to teleport back using a `null` chronosphere.

pchote replied:
My latest rewrite may have regressed the lua handling - will look into it.

pchote commented:
Fixed. Actor references can now be specified as either an actor or a string map name, and actors can now return without a valid chronosphere.

abcdefg30 reviewed (approved):
Lgtm.

teinarss commented:
Code looks good, havent had time to test it yet.

teinarss commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/908442391205a0c2ca4d539d30b2417d854f64fb...269a8fee6732e9091549592118ae415bb77bda4c

atlimit8 commented on OpenRA.Game/Map/ActorInitializer.cs:
```diff
@@ -53,59 +65,178 @@ public ActorInitializer(Actor actor, TypeDictionary dict)
+		public virtual void Initialize(T value)
+		{
+			var field = GetType().GetField("value", BindingFlags.NonPublic | BindingFlags.Instance);
+			if (field != null)
```
Since `GetType()` will return the reflection of this class or a subclass with the `protected readonly T value` field, when can `field` be null? A louder failure might be better here.
